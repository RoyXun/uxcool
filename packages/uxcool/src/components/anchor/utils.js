const REG_ANCHOR_ID = /#[^#]+$/;

export function getAnchorRelateElement(href = '') {
  const matches = href.match(REG_ANCHOR_ID);
  if (!matches) {
    return null;
  }

  return document.querySelector(matches[0]);
}

export function getElementOffsetTop(element, container) {
  let top = 0;
  if (!element || !container) {
    return top;
  }

  const box = element.getBoundingClientRect();
  if (container.window === window) {
    const doc = element.ownerDocument;
    const docElement = doc && doc.documentElement;
    const body = doc && doc.body;
    top = box.top - docElement.clientTop || body.clientTop || 0;
  } else {
    const containerRect = container.getBoundingClientRect();
    top = box.top - containerRect.top;
  }
  return top;
}
