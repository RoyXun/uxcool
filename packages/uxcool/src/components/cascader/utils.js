import { isFunction } from '@suning/v-utils';

export const SPLIT_STR = '\u0001';

export const DEFAULT_FIELD_NAMES = {
  label: 'label',
  value: 'value',
  children: 'children',
};

export function genCascaderKey(valuePath) {
  return valuePath.join(SPLIT_STR);
}

export function delay(fn, wait) {
  if (!isFunction(fn)) {
    return null;
  }
  return setTimeout(fn, wait);
}
