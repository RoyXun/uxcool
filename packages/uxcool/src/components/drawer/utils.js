import { isNaN, isNumber } from '@cloud-sn/v-utils';

export function genUid() {
  let dt = new Date().getTime();
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    // eslint-disable-next-line no-bitwise, no-mixed-operators
    const r = ((dt + Math.random() * 16) % 16) | 0;
    dt = Math.floor(dt / 16);
    // eslint-disable-next-line no-bitwise,
    return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
  });
}

export function noop() {}

export function isNumeric(value) {
  return !isNaN(value) && isNumber(value);
}

export function hasScrollBar() {
  return document.body.clientWidth < window.innerWidth;
}

export function getVisibleDrawerCnt(drawers = {}) {
  return Object.keys(drawers).filter((k) => drawers[k]).length;
}

export function hasVisibleDrawer(drawers = {}) {
  return getVisibleDrawerCnt(drawers) >= 1;
}

export function getDoc() {
  return window ? window.document.body : {};
}
