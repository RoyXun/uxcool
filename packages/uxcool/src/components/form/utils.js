import omit from 'object.omit';
import {
  warning,
  toArray,
  isFunction,
  isPlainObject,
  isString,
  setValueByPath,
  getValueByPath,
  isDef,
} from '@suning/v-utils';

import { Validator } from 'vee-validate';

const { hasOwnProperty } = Object.prototype;

export function isValidateField(field) {
  return field && isDef(field.rules);
}

export function getSlotOrValue(name, target) {
  const { $slots } = target;
  if ($slots[name]) {
    return $slots[name];
  }

  return target[name] || null;
}

export function hasPath(path, target) {
  let obj = target;
  return path.split('.').every((prop) => {
    if (!hasOwnProperty.call(obj, prop)) {
      return false;
    }
    obj = obj[prop];

    return true;
  });
}

// export const getPath = (path, target, def) => {
//   if (!path || !target) return def;

//   let value = target;
//   const paths = path.split('.');

//   for (let i = 0, l = paths.length; i < l; i += 1) {
//     const prop = paths[i];
//     if (!hasOwnProperty.call(value, prop) && value[prop] === undefined) {
//       value = def;
//       break;
//     }

//     value = value[prop];
//   }
//   return value;
// };

export function getPath(path, target, def) {
  return getValueByPath(target, path, def);
}
export function find(arrayLike, predicate) {
  const array = Array.isArray(arrayLike) ? arrayLike : toArray(arrayLike);
  for (let i = 0; i < array.length; i += 1) {
    if (predicate(array[i])) {
      return array[i];
    }
  }

  return undefined;
}

export function getModel(vnode) {
  return vnode.data
    ? vnode.data.model || find(vnode.data.directives, d => d.name === 'model')
    : null;
}

export function getVnodeOptions(vnode) {
  const component = vnode.componentInstance;
  const opts = {};
  return component
    ? {
      model: component.$options.model || {},
      listeners: component.$listeners || {},
    }
    : opts;
}

export function resolveModel(vnode) {
  const model = getModel(vnode);
  if (!model) {
    return null;
  }

  // https://github.com/vuejs/vue/blob/dev/src/core/util/lang.js#L26
  const watchable = !/[^\w.$]/.test(model.expression) && hasPath(model.expression, vnode.context);
  const lazy = !!(model.modifiers && model.modifiers.lazy);
  if (!watchable) {
    return {
      expression: null,
      lazy,
    };
  }

  return {
    expression: model.expression,
    lazy,
  };
}

export function resolveInitialValue(vnode) {
  const model = getModel(vnode);

  return model && model.value;
}

export function makeVM(vm, context) {
  return {
    get $el() {
      return vm.$el;
    },
    get $refs() {
      // 同理
      return context.$refs;
    },
    get context() {
      return context;
    },
    // $watch会观察model的expression,所以要用vnode.context
    $watch: context.$watch ? context.$watch.bind(context) : () => {},
    $validator: vm.$validator
      ? {
        errors: vm.$validator.errors,
        validate: vm.$validator.validate.bind(vm.$validator),
        update: vm.$validator.update.bind(vm.$validator),
      }
      : null,
  };
}

export function getCtorConfig(vnode) {
  if (!vnode.componentInstance) return null;

  const config = getPath('componentInstance.$options.$_veeValidate', vnode);

  return config;
}

export function resolveSetter(name, el, vnode) {
  const model = getModel(vnode);
  if (model && model.expression) {
    if (model.callback) {
      return model.callback;
    }

    if (model.expression && hasPath(vnode.context, model.expression)) {
      return val => setValueByPath(vnode.context, model.expression, val);
    }
  }

  const options = getVnodeOptions(vnode);
  const eventName = options.model.event || 'input';
  const listener = options.listeners[eventName];
  if (listener) {
    return val => vnode.componentInstance.$emit(eventName, val);
  }

  return () => {
    warning(
      false,
      `Please provide the \`v-model\` property or bind \`${eventName}\` events for the elements within the \`${name}\` FieldDecorator.`
    );
  };
}
export function resolveGetter(el, vnode, model, valuePath) {
  if (model && model.expression) {
    return () => getPath(model.expression, vnode.context);
  }

  if (valuePath) {
    return () => getPath(valuePath, vnode.context);
  }

  if (vnode.componentInstance) {
    const config = getCtorConfig(vnode);
    if (config && isFunction(config.value)) {
      const boundGetter = config.value.bind(vnode.componentInstance);

      return () => boundGetter();
    }

    return () => vnode.componentInstance.value;
  }

  switch (el.type) {
    case 'checkbox':
      return () => {
        let els = document.querySelectorAll(`input[name="${el.name}"]`);

        els = toArray(els).filter(v => v.checked);
        if (!els.length) return undefined;

        return els.map(checkbox => checkbox.value);
      };
    case 'radio':
      return () => {
        const els = document.querySelectorAll(`input[name="${el.name}"]`);
        const elm = find(els, v => v.checked);

        return elm && elm.value;
      };
    case 'file':
      return () => toArray(el.files);
    case 'select-multiple':
      return () =>
        toArray(el.options)
          .filter(opt => opt.selected)
          .map(opt => opt.value);
    default:
      return () => el && el.value;
  }
}

function parseRule(rule) {
  let params = [];
  const name = rule.split(':')[0];

  if (rule.indexOf(':') > -1) {
    params = rule
      .split(':')
      .slice(1)
      .join(':')
      .split(',');
    // 处理confirmed, after, before等规则
    if (Validator.isTargetRule(name)) {
      const selector = params[0];
      if (selector) {
        // @开头的selector 标示用户使用元素自定义属性
        // 由于元素存在嵌套的情况veeValidate的updateDependencies不能取到实际元素,导致无法获取实际值,所以让用户指定实际值
        if (selector.indexOf('@') !== 0) {
          params[0] = `$${selector.replace(/^\$/, '')}Ref`;
        } else {
          params[0] = selector.slice(1);
        }
      }
    }
  }
  return { name, params };
}
export function parseStringRules(rules) {
  return rules.split('|').reduce((p, rule) => {
    const np = p;
    const parsedRule = parseRule(rule);
    if (!parsedRule.name) {
      return np;
    }

    np[parsedRule.name] = parsedRule.params;
    return np;
  }, {});
}
export function parseFieldRules(rules) {
  let r = {
    rules: {},
    messages: {},
  };

  if (Array.isArray(rules)) {
    r = rules.reduce((p, v) => {
      const np = p;
      const rule = parseFieldRules(v);

      np.rules = { ...np.rules, ...rule.rules };
      np.messages = { ...np.messages, ...rule.messages };

      return np;
    }, r);
  } else if (isPlainObject(rules)) {
    if ('message' in rules) {
      const nRules = omit(rules, 'message');
      const keys = Object.keys(nRules);
      if (keys.length >= 1) {
        r = {
          rules: nRules,
          messages: {
            [keys[0]]: rules.message,
          },
        };
      }
    } else if (!('message' in rules)) {
      r.rules = rules;
    }
  } else if (isString(rules)) {
    r.rules = parseStringRules(rules);
  }

  return r;
}

export function validateField(field, validator) {
  if (field.initial) {
    validator.validate(`#${field.id}`, field.value);
  } else {
    // eslint-disable-next-line
    validator._validate(field, field.value, true).then(result => {
      field.setFlags({
        valid: result.valid,
        invalid: !result.valid,
      });
    });
  }
}
