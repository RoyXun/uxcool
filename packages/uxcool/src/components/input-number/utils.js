import isNaN from 'lodash/isNaN';

export { isNaN };

export function isNotCompleteNumber(v) {
  return (
    isNaN(Number(v)) ||
    v === '' ||
    v === null ||
    v === undefined ||
    (v && v.toString().slice('.') === v.length - 1)
  );
}

export function getPrecision(val, defaultPrecision = null) {
  if (defaultPrecision !== null) {
    return defaultPrecision;
  }

  let valStr = val.toString();
  let eDigit = 0;
  if (/e-/.test(valStr)) {
    [valStr, eDigit] = valStr.split('e-');
  }

  let decimal = 0;
  if (/\./.test(valStr)) {
    decimal = parseInt(valStr.split('.')[1].length, 10);
  }

  return decimal + eDigit;
}

export function getMaxPrecision(val, step, ratio = 1, defaultPrecision = null) {
  if (defaultPrecision !== null) {
    return defaultPrecision;
  }

  const stepPrecision = getPrecision(step);
  const ratioPrecision = getPrecision(ratio);
  const valPrecision = getPrecision(val);

  return Math.max(valPrecision, stepPrecision + ratioPrecision);
}

export function getRatio(e) {
  let ratio = 1;
  if (e.metaKey || e.ctrlKey) {
    ratio = 0.1;
  } else if (e.shiftKey) {
    ratio = 10;
  }
  return ratio;
}
