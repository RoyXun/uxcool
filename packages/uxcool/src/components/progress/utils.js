export function defaultInfoFormat(percent) {
  return `${percent}%`;
}

export function getCirclePath(pos, radius) {
  const { start, end } = pos;
  return `M 50 50 m ${start.x},${start.y}
          a ${radius},${radius} 0 1 1 ${end.x},${end.y}
          a ${radius},${radius} 0 1 1 ${-1 * end.x},${-1 * end.y}`;
}

export function getPosition(gapPos, radius) {
  let start = {
    x: 0,
    y: -radius,
  };
  let end = {
    x: 0,
    y: 2 * radius,
  };

  switch (gapPos) {
    case 'left':
      start = {
        x: -radius,
        y: 0,
      };
      end = {
        x: 2 * radius,
        y: 0,
      };
      break;
    case 'right':
      start = { x: radius, y: 0 };
      end = {
        x: -2 * radius,
        y: 0,
      };
      break;
    case 'bottom':
      start = {
        x: 0,
        y: radius,
      };
      end = {
        x: 0,
        y: -2 * radius,
      };
      break;
    default:
      break;
  }

  return {
    start,
    end,
  };
}
