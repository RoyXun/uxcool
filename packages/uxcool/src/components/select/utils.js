import {
  isSameTypeVNode,
  isArray,
  cloneVNode,
  isDef,
  isFunction,
  warning,
  extractVNodeData,
  isVueComponent,
} from '@cloud-sn/v-utils';

export function isValidValue(value) {
  return isDef(value) && value !== '';
}
export function buildOptionsFromSlot(nodes = []) {
  const ret = [];

  for (let i = 0, l = nodes.length; i < l; i += 1) {
    const node = nodes[i];
    if (isSameTypeVNode(node, 'isOptionGroupType')) {
      const { componentOptions: cp, data: { attrs } } = node;
      const props = cp.propsData;
      if (process.env.NODE_ENV !== 'production' && !isDef(props.label)) {
        warning(
          false,
          'OptionGroup invalid prop: type check failed for prop "label". Expected String,got undefined .'
        );
      }
      ret.push({
        ...attrs,
        ...props,
        label: props.label,
        children: buildOptionsFromSlot(cp.children),
      });
    } else if (isSameTypeVNode(node, 'isOptionType')) {
      const { componentOptions: cp, data: { attrs } } = node;
      const props = cp.propsData;
      if (process.env.NODE_ENV !== 'production' && !isDef(props.value)) {
        warning(
          false,
          'Option invalid prop: type check failed for prop "value". Expected String, Number ,got undefined .'
        );
      }
      ret.push({
        ...attrs,
        ...props,
        value: props.value,
        label: props.label || '',
        labelNode: cp.children,
        disabled: 'disabled' in props && props.disabled !== false,
      });
    }
  }
  return ret;
}

export function isUnRenderOption(option) {
  return 'UNRENDER_OTPION' in option && option.UNRENDER_OTPION === true;
}
export function isOptionGroup(option) {
  return 'children' in option;
}

export function cloneLabelNodes(vnodes) {
  return isArray(vnodes) ? vnodes.map((node) => cloneVNode(node)) : cloneVNode(vnodes);
}

export function normalizeContent(maxLen, content) {
  const l = Number(maxLen);
  if (l <= 0 || typeof content !== 'string' || content.length < l) {
    return content;
  }

  return `${content.slice(0, l)}...`;
}

export function DefaultFilterOption(filterProp) {
  return (inputValue, option) => {
    if (option.disabled) {
      return false;
    }
    const optionValue = String(option[filterProp]) || '';
    return optionValue.toLowerCase().indexOf(inputValue.toLowerCase()) > -1;
  };
}

export function recursiveSearch(options = [], inputValue, filterFn) {
  const ret = [];
  for (let i = 0, l = options.length; i < l; i += 1) {
    const option = options[i];
    if (isOptionGroup(option)) {
      const { children } = option;
      const nChildren = recursiveSearch(children || [], inputValue, filterFn);
      if (nChildren.length > 0) {
        ret.push({ ...option, children: nChildren });
      }
    } else if (filterFn(inputValue, option.originNode)) {
      ret.push(option);
    }
  }
  return ret;
}

export function buildOptionOriginNode(name, optionMap) {
  if (!isDef(name)) {
    return {};
  }
  if (name in optionMap) {
    const { originNode } = optionMap[name];
    const {
      label, value, title, disabled
    } = originNode;
    const nLabel = label || value;
    return {
      ...originNode,
      disabled: !!disabled,
      title: title || nLabel,
      label: nLabel,
      value,
    };
  }
  return {
    disabled: false,
    title: name,
    label: name,
    value: name,
  };
}

export function calcMultipleSearchInputWidth(inputValue, inputMirror) {
  if (inputValue && inputMirror) {
    return {
      width: `${inputMirror.clientWidth}px`,
    };
  }
  return {};
}

function getRegByTokens(tokens = []) {
  return new RegExp(`^[^${tokens.join('')}]+[${tokens.join('')}]`);
}
export function includeSeparators(value = '', tokens = []) {
  const reg = getRegByTokens(tokens);
  return reg.test(value);
}

export function getOptionLabelMap(optionMap = {}) {
  return Object.keys(optionMap).reduce((r, k) => {
    const nr = r;
    const option = optionMap[k];
    nr[option.label || k] = option;
    return nr;
  }, {});
}

export function splitValueBySeparator(value = '', tokens = []) {
  const splitReg = new RegExp(`[${tokens.join('')}]`);
  return value.split(splitReg).filter((v) => !!v);
}

export function getComboboxValue(value = '', optionMap = {}, optionLabelProp = 'value') {
  const comboboxValue = isArray(value) ? value[0] : value;
  let label = '';
  if (comboboxValue) {
    if (optionLabelProp === 'value') {
      label = comboboxValue;
    } else {
      // optionLabelProp = label;
      const option = optionMap[comboboxValue];
      if (option) {
        label = option.label; //eslint-disable-line
      } else {
        label = comboboxValue;
      }
    }
  }
  return label;
}

export function getOptionOriginNode(optionMap = {}) {
  return Object.keys(optionMap).map((k) => {
    const option = optionMap[k];
    return option.originNode;
  });
}

export function genInputElement(inputNode, data = {}) {
  const overlayData = extractVNodeData(inputNode, isVueComponent(inputNode));

  const { input: originInput } = overlayData.on || {};
  const { input: onInput } = data.on || {};
  const inputFn = isFunction(onInput) ? (_, e) => onInput(e) : () => {};
  const inputInput = originInput ? [originInput, inputFn] : inputFn;
  const ref = data.ref || inputNode.data.ref;
  return cloneVNode(
    inputNode,
    {
      class: [overlayData.class, data.class],
      props: {
        ...overlayData.props,
        ...data.props,
      },
      on: {
        ...overlayData.on,
        input: inputInput,
      },
      ref,
    },
    false
  );
}
