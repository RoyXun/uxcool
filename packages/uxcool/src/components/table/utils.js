import { warning } from '../utils';

export function getColKey(col, colIdx) {
  return col.key || `$$_id${colIdx}`;
}
export function getRowKey(rowKey, record, rowIdx) {
  const key = typeof rowKey === 'function' ? rowKey(record, rowIdx) : record[rowKey];
  warning(
    key !== undefined,
    'Each record in dataSource of table should have a unique `key` prop, or set `rowKey` to an unique primary key'
  );
  return key === undefined ? `$$_id${rowIdx}` : String(key);
}

export function isFunction(fn) {
  return typeof fn === 'function';
}

export function flatRows(rows = [], childColName = 'children', deep = true) {
  return rows.reduce((r, v) => {
    const nr = r;
    nr.push(deep ? { ...v } : v);
    const child = v[childColName];
    if (Array.isArray(child) && child.length > 0) {
      nr.push(...flatRows(v[childColName], childColName, deep));
    }
    return nr;
  }, []);
}

function recursiveRows(rows = [], cb, childColName, deep = true) {
  const ret = [];
  for (let i = 0, l = rows.length; i < l; i += 1) {
    const row = rows[i];
    const r = deep ? { ...row } : row;
    ret.push(isFunction(cb) ? cb(r) : r);
    const child = row[childColName];
    if (Array.isArray(child) && child.length > 0) {
      r[childColName] = recursiveRows(child, cb, childColName, deep);
    }
  }
  return ret;
}

export function normalizeRows(rows = [], cb = null, childColName = 'children', deep = true) {
  return recursiveRows(rows, cb, childColName, deep);
}

export function normalizeCols(cols = [], cb = null, childColName = 'children', deep = true) {
  return recursiveRows(cols, cb, childColName, deep);
}

export function recursiveSort(data = [], sortFn, childColName = 'children') {
  return data.sort(sortFn).map((v) => {
    const nv = v;
    const child = nv[childColName];
    if (Array.isArray(child) && child.length > 0) {
      nv[childColName] = recursiveSort(child, sortFn, childColName);
    }
    return nv;
  });
}
