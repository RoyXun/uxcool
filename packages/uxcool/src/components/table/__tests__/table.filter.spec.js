import {
  mountPickerFactory, triggerEvent, waitTime, mount
} from '@cloud-sn/v-test-utils';
import { getData, getFilterDropdownPortal } from './utils';
import Table from '../table';
import FilterDropdown from '../filterDropdown';

describe('Table.filter', () => {
  const mountPicker = mountPickerFactory(Table);
  const column = {
    key: 'name',
    title: 'Name',
    dataIndex: 'name',
    filters: [
      {
        text: 'A1',
        value: 'a1',
      },
      {
        text: 'B1',
        value: 'b1',
      },
      {
        text: 'C1',
        value: 'c1',
        children: [
          {
            text: 'CC1',
            value: 'cc1',
          },
          {
            text: 'CC2',
            value: 'cc2',
          },
        ],
      },
    ],
    onFilter(val, record) {
      return record.name.includes(val);
    },
  };
  const dataSource = getData();
  it('render filter', async () => {
    const wrapper = await mountPicker({
      propsData: {
        value: dataSource,
        columns: [column],
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('render filter dropdown', async () => {
    const wrapper = await mountPicker({
      propsData: {
        value: dataSource,
        columns: [column],
      },
    });

    const portal = await getFilterDropdownPortal(wrapper);
    expect(portal.find('.ux-dropdown-content').html()).toMatchSnapshot();
  });

  it('render radio filter ', async () => {
    const wrapper = await mountPicker({
      propsData: {
        value: dataSource,
        columns: [{ ...column, filterMultiple: false }],
      },
    });

    const portal = await getFilterDropdownPortal(wrapper);
    expect(portal.find('.ux-dropdown-content').html()).toMatchSnapshot();
  });

  it('render custom dropdown', async () => {
    const wrapper = await mountPicker({
      propsData: {
        value: dataSource,
        columns: [
          {
            ...column,
            // eslint-disable-next-line
            filterDropdown(h) {
              return <div class="custom-filter"> this is custome filter </div>;
            },
          },
        ],
      },
    });

    const portal = await getFilterDropdownPortal(wrapper);
    expect(portal.find('.ux-dropdown-content').html()).toMatchSnapshot();
  });

  it('controlled filterDropdownVisible', async () => {
    const Test = {
      props: {
        filterDropdownVisible: {
          type: Boolean,
          default: false,
        },
      },
      render() {
        const { filterDropdownVisible } = this;
        const props = {
          value: dataSource,
          columns: [{ ...column, filterDropdownVisible }],
        };
        return <Table {...{ props }} />;
      },
    };
    const wrapper = mount(Test, {
      propsData: {
        filterDropdownVisible: true,
      },
    });
    await waitTime();
    const filterDropdown = wrapper.find(FilterDropdown);
    expect(filterDropdown.vm.dropdownVisible).toBeTruthy();

    wrapper.setProps({
      filterDropdownVisible: false,
    });
    await waitTime();
    expect(filterDropdown.vm.dropdownVisible).toBeFalsy();
  });

  it('filter keep selected keys when table value updated', async () => {
    const wrapper = await mountPicker({
      propsData: {
        value: dataSource,
        columns: [{ ...column, filterDropdownVisible: true }],
      },
    });

    const filterDropdown = wrapper.find(FilterDropdown);
    expect(filterDropdown.vm.innerSelectedKeys).toEqual([]);

    const portal = await getFilterDropdownPortal(wrapper);

    const checkboxs = portal.findAll('.ux-dropdown-menu-item');

    await triggerEvent(checkboxs.at(0), 'click', 100);
    await triggerEvent(portal.find('.ux-table-filter-dropdown-link.confirm'), 'click');

    expect(filterDropdown.vm.innerSelectedKeys).toEqual(['a1']);

    wrapper.setProps({
      value: [...dataSource, { key: '4', name: 'test' }],
    });
    await waitTime();

    expect(filterDropdown.vm.innerSelectedKeys).toEqual(['a1']);
  });

  it('onFilterDropdownVisibleChange', async () => {
    const onFilterDropdownVisibleChange = jest.fn();
    const wrapper = await mountPicker({
      propsData: {
        value: dataSource,
        columns: [{ ...column, onFilterDropdownVisibleChange }],
      },
    });

    await getFilterDropdownPortal(wrapper);

    expect(onFilterDropdownVisibleChange).toHaveBeenCalledWith(true);
  });

  describe('filteredValue', () => {
    it.each([
      ['[]', 'John', []],
      ['null', 'John', null],
    ])('controlled filteredValue to %s', async (_, v1, v2) => {
      const wrapper = await mountPicker({
        propsData: {
          value: dataSource,
          columns: [{ ...column, filteredValue: [v1] }],
        },
      });

      expect(wrapper.findAll('.ux-table-tbody > .ux-table-row').length).toBe(1);
      wrapper.setProps({
        columns: [{ ...column, filteredValue: v2 }],
      });
      await waitTime();

      expect(wrapper.findAll('.ux-table-tbody > .ux-table-row').length).toBe(dataSource.length);
    });
  });

  describe('change', () => {
    it('fire change', async () => {
      const onChange = jest.fn();
      const wrapper = await mountPicker({
        propsData: {
          value: dataSource,
          columns: [column],
        },
        listeners: {
          change: onChange,
        },
      });

      const portal = await getFilterDropdownPortal(wrapper);
      const checkboxs = portal.findAll('.ux-dropdown-menu-item');

      await triggerEvent(checkboxs.at(0), 'click', 100);
      await triggerEvent(portal.find('.ux-table-filter-dropdown-link.confirm'), 'click');

      expect(onChange).toHaveBeenCalled();
      expect(onChange.mock.calls[0][1]).toEqual({ name: ['a1'] });
    });
    it('should not fire change without change anything', async () => {
      const onChange = jest.fn();
      const wrapper = await mountPicker({
        propsData: {
          value: dataSource,
          columns: [column],
        },
        listeners: {
          change: onChange,
        },
      });
      const portal = await getFilterDropdownPortal(wrapper);
      await triggerEvent(portal.find('.ux-table-filter-dropdown-link.clear'), 'click');
      expect(onChange).not.toHaveBeenCalled();
    });

    it('fire change when dropdown hidden', async () => {
      const onChange = jest.fn();
      const wrapper = await mountPicker({
        propsData: {
          value: dataSource,
          columns: [column],
        },
        listeners: {
          change: onChange,
        },
      });
      const portal = await getFilterDropdownPortal(wrapper);

      const checkboxs = portal.findAll('.ux-dropdown-menu-item');
      await triggerEvent(checkboxs.at(0), 'click', 100);
      await triggerEvent(wrapper.find('.ux-table-filter-icon'), 'click');

      expect(onChange).toHaveBeenCalled();
      expect(onChange.mock.calls[0][1]).toEqual({ name: ['a1'] });
    });
  });

  it('custom filter icon', async () => {
    const wrapper = await mountPicker({
      propsData: {
        value: dataSource,
        columns: [
          {
            ...column,
            // eslint-disable-next-line
            filterIcon(h) {
              return <span>filter</span>;
            },
          },
        ],
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('pagination', () => {
    const pData = [];
    for (let i = 0; i < 30; i += 1) {
      pData.push({
        key: i,
        name: `a1-name-${i}`,
      });
    }
    it('should reset pagination after filter', async () => {
      const onChange = jest.fn();
      const wrapper = await mountPicker({
        propsData: {
          value: pData,
          columns: [column],
          pagination: true,
        },
        listeners: {
          change: onChange,
        },
      });

      const portal = await getFilterDropdownPortal(wrapper);
      const checkboxs = portal.findAll('.ux-dropdown-menu-item');
      await triggerEvent(checkboxs.at(0), 'click', 100);
      await triggerEvent(portal.find('.ux-table-filter-dropdown-link.confirm'), 'click');

      expect(onChange).toHaveBeenCalled();
      expect(onChange.mock.calls[0][0]).toEqual(
        expect.objectContaining({
          current: 1,
          pageSize: 10,
        })
      );
      expect(onChange.mock.calls[0][1]).toEqual({ name: ['a1'] });
    });
  });
});
