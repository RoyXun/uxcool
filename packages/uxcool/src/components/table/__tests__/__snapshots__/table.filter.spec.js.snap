// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.filter custom filter icon 1`] = `
<div class="ux-table-wrapper ux-table-light">
  <div class="ux-spin-nested-loading">
    <div class="ux-spin" style="display: none;"><svg viewBox="25 25 50 50" class="ux-spin-circle">
        <circle cx="50" cy="50" r="20" fill="none"></circle>
      </svg> </div>
    <div class="ux-spin-container">
      <div class="ux-table ux-table-scroll-position-left ux-table-default">
        <div class="ux-table-content">
          <div class="ux-table-main-table">
            <!---->
            <div class="ux-table-body">
              <table class="">
                <colgroup>
                  <col>
                </colgroup>
                <thead class="ux-table-thead">
                  <tr>
                    <th class="ux-table-column-has-filters"><span>Name<span title="筛选" class="ux-table-filter-icon ux-dropdown-trigger"><span>filter</span></span></span></th>
                  </tr>
                </thead>
                <tbody class="ux-table-tbody">
                  <tr class="ux-table-row">
                    <td class="ux-table-column-has-filters">John Brown</td>
                  </tr>
                  <tr class="ux-table-row">
                    <td class="ux-table-column-has-filters">Jim Green</td>
                  </tr>
                  <tr class="ux-table-row">
                    <td class="ux-table-column-has-filters">Joe Black</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <!---->
            <!---->
          </div>
        </div>
        <div class="ux-table-resize-line" style="display: none;"></div>
      </div>
    </div>
  </div>
  <!---->
</div>
`;

exports[`Table.filter render custom dropdown 1`] = `
<div class="ux-dropdown-content">
  <div>
    <div class="custom-filter"> this is custome filter </div>
  </div>
</div>
`;

exports[`Table.filter render filter 1`] = `
<div class="ux-table-wrapper ux-table-light">
  <div class="ux-spin-nested-loading">
    <div class="ux-spin" style="display: none;"><svg viewBox="25 25 50 50" class="ux-spin-circle">
        <circle cx="50" cy="50" r="20" fill="none"></circle>
      </svg> </div>
    <div class="ux-spin-container">
      <div class="ux-table ux-table-scroll-position-left ux-table-default">
        <div class="ux-table-content">
          <div class="ux-table-main-table">
            <!---->
            <div class="ux-table-body">
              <table class="">
                <colgroup>
                  <col>
                </colgroup>
                <thead class="ux-table-thead">
                  <tr>
                    <th class="ux-table-column-has-filters"><span>Name<span title="筛选" class="ux-table-filter-icon ux-dropdown-trigger"><i class="fu fu-filter"></i></span></span></th>
                  </tr>
                </thead>
                <tbody class="ux-table-tbody">
                  <tr class="ux-table-row">
                    <td class="ux-table-column-has-filters">John Brown</td>
                  </tr>
                  <tr class="ux-table-row">
                    <td class="ux-table-column-has-filters">Jim Green</td>
                  </tr>
                  <tr class="ux-table-row">
                    <td class="ux-table-column-has-filters">Joe Black</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <!---->
            <!---->
          </div>
        </div>
        <div class="ux-table-resize-line" style="display: none;"></div>
      </div>
    </div>
  </div>
  <!---->
</div>
`;

exports[`Table.filter render filter dropdown 1`] = `
<div class="ux-dropdown-content">
  <div class="ux-table-filter-dropdown">
    <ul role="menu" class="ux-dropdown-menu ux-dropdown-menu-root ux-dropdown-menu-light ux-dropdown-menu-vertical">
      <li title="" class="ux-dropdown-menu-item"><label class="ux-checkbox-wrapper"><span class="ux-checkbox"><input type="checkbox" class="ux-checkbox-input" value=""> <span class="ux-checkbox-inner"></span></span>
          <!----></label><span>A1</span></li>
      <li title="" class="ux-dropdown-menu-item"><label class="ux-checkbox-wrapper"><span class="ux-checkbox"><input type="checkbox" class="ux-checkbox-input" value=""> <span class="ux-checkbox-inner"></span></span>
          <!----></label><span>B1</span></li>
      <li class="ux-dropdown-menu-submenu ux-dropdown-menu-submenu-vertical">
        <div title="" class="ux-dropdown-menu-submenu-title"><span>C1</span></div>
      </li>
    </ul>
    <div class="ux-table-filter-dropdown-btns"><a class="ux-table-filter-dropdown-link confirm">确定</a><a class="ux-table-filter-dropdown-link clear">重置</a></div>
  </div>
</div>
`;

exports[`Table.filter render radio filter  1`] = `
<div class="ux-dropdown-content">
  <div class="ux-table-filter-dropdown">
    <ul role="menu" class="ux-dropdown-menu ux-dropdown-menu-root ux-dropdown-menu-light ux-dropdown-menu-vertical">
      <li title="" class="ux-dropdown-menu-item"><label class="ux-radio-wrapper"><span class="ux-radio"><input type="radio" name="" class="ux-radio-input" value=""> <span class="ux-radio-inner"></span></span>
          <!----></label><span>A1</span></li>
      <li title="" class="ux-dropdown-menu-item"><label class="ux-radio-wrapper"><span class="ux-radio"><input type="radio" name="" class="ux-radio-input" value=""> <span class="ux-radio-inner"></span></span>
          <!----></label><span>B1</span></li>
      <li class="ux-dropdown-menu-submenu ux-dropdown-menu-submenu-vertical">
        <div title="" class="ux-dropdown-menu-submenu-title"><span>C1</span></div>
      </li>
    </ul>
    <div class="ux-table-filter-dropdown-btns"><a class="ux-table-filter-dropdown-link confirm">确定</a><a class="ux-table-filter-dropdown-link clear">重置</a></div>
  </div>
</div>
`;
